_posts/2023-06-22-a-simple-rest-api-with-play-framework-and-scala.md

```diff
@@ -0,0 +1,420 @@
+
+```scala
+package persistence
+
+trait Persistence[A, B, C[_]]{
+  val db: C[A]
+  def insert(entity: A): A
+  def find(id: B): Option[A]
+  def update(entity: A): Unit
+  def delete(id: B): Unit
+  def findAll(): List[A]
+}
+```
+
+Our persisted entity for this model we use a simple Car object:
+
+```scala
+package persistence.entities
+
+case class Car(id: Long, brand: String, model: String, cc: Int)
+```
+
+and for an implementation for our persistence mechanism we use an in memory singleton object
+
+```scala
+package persistence
+
+import persistence.entities.Car
+
+import scala.collection.mutable.ListBuffer
+import scala.util.Random
+
+object InMemoryPersistence extends Persistence[Car, Long, ListBuffer]{
+  override val db: ListBuffer[Car] = ListBuffer.empty
+  val idGenerator = Random
+
+  def insert(car: Car): Car = {
+    val idx = db.indexWhere(_.id == car.id)
+    val newCar = car.copy(id=idGenerator.nextLong())
+    if(idx > -1) throw new RuntimeException(s"Duplicate element with id: ${car.id}")
+    else db.addOne(newCar)
+
+    newCar
+  }
+  def find(id: Long): Option[Car] = db.find(_.id == id)
+
+  def update(car: Car): Unit = {
+    val idx = db.indexWhere(_.id == car.id)
+    if(idx > -1) db.update(idx, car)
+    else throw new RuntimeException(s"No element to update with id: ${car.id}")
+  }
+
+  def delete(id: Long): Unit = {
+    val idx = db.indexWhere(_.id == id)
+    if(idx > -1 ) db.remove(idx)
+    else throw new RuntimeException(s"No element to delete with id: ${id}")
+  }
+
+  def findAll(): List[Car] = db.toList
+}
+```
+
+Close to our persistence mechanism we define our repositories, responsible for handling all the specifics for saving, retrieving etc.
+
+```scala
+package repositories
+
+import persistence.Persistence
+import persistence.entities.Car
+import repositories.errors.RepositoryError
+
+import scala.collection.mutable.ListBuffer
+
+class CarRepository(val db: Persistence[Car, Long, ListBuffer]){
+  def findById(id: Long): Either[RepositoryError, Option[Car]] = handleIfErrors(db.find(id))
+  def save(car: Car): Either[RepositoryError, Car] = handleIfErrors(db.insert(car))
+  def update(car: Car): Either[RepositoryError, Unit] = handleIfErrors(db.update(car))
+  def delete(id: Long): Either[RepositoryError, Unit] = handleIfErrors(db.delete(id))
+  def findAll(): Either[RepositoryError, List[Car]] = handleIfErrors(db.findAll())
+
+  private def handleIfErrors[A](f: => A) = {
+    try{
+      Right(f)
+    } catch {
+      case e: RuntimeException => Left(RepositoryError(e.getMessage))
+    }
+  }
+}
+```
+## 3.2 Managers
+On this layer we define our managers. Usually containing logic to manage other resources like repositories, transactions, connections, security etc and contains business logic (here is too simple) of our application.
```

Comment:
Need to expand this explanation a bit. You may want to give an example. You can get this to say "the example in the code below is very simple because we simply invoke methods from the repository. But assume we also wanted to keep track of the most popular cars that were requested (so that the marketing department can do their thing with the data), so we'd do some more than just fetching the cars from the database, we'd do ........"